Adding support for ivshmem

tbursztyka opened the pull request:
Adding support for Inter-VM Shared Memory device (ivshmem).
Such device can be brought by Qemu or ACRN in order to share memory between the VM and the host or between VMs.

Commit:
drivers/virtualization: Adding ivshmem driver

This is placed into drivers/virtualization as it does not belong to any
existing subsystem.

This is only the ivshmem-plain variant.

This device is provided by qemu or ACRN, and can be used to share memory
either between the host and the VM or between VMs. Here if zephyr is
used as a VM, it will be able to take advantage of such feature.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/virtualization: Add doorbell support for ivshmem driver

This basically adds support for an interrupt based ivshmem variant
called "ivshmem-doorbell".

This allows, via MSI-X, to get multiple vectors for notifications, get
assigned and ID and being able to send a message to another ID (another
VM).

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
@jhedberg @dcpleung @andrewboie @andyross  please review, I'll provide a sample and/or test next week. But so far testing against ivshmem-sever/client worked pretty well.

Commit:
drivers/virtualization: Add a shell module for the ivshmem driver

This easily permits to test ivshmem and basic functionality of a custom
protocol passing through.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

jakub-uC commented on drivers/misc/Kconfig:
```diff
@@ -0,0 +1,63 @@
+	  Interrupt priority used for the MSI-X generated interrupts.
+
+config IVSHMEM_SHELL
+	bool "Enable IVshmem shell module"
```
Is it a good idea to put here:
`depends on SHELL`?

tbursztyka replied:
yes

jakub-uC commented on drivers/misc/misc_ivshmem_shell.c:
```diff
@@ -0,0 +1,93 @@
+	if (ivshmem == NULL) {
+		ivshmem = device_get_binding(CONFIG_IVSHMEM_DEV_NAME);
+		if (!ivshmem) {
+			shell_error(shell, "IVshmem device cannot be found\n");
```
The macro `shell_error` already adds a new line. If you want to have 2 lines break it is fine.

tbursztyka replied:
no indeed, it's an error. probably got mixed up with shell_fprintf

jakub-uC commented on drivers/misc/misc_ivshmem_shell.c:
```diff
@@ -0,0 +1,175 @@
+SHELL_CMD_REGISTER(ivshmem, &sub_ivshmem_cmds,
+		   "IVshmem information", cmd_ivshmem_shmem);
+
+SHELL_CMD_REGISTER(ivshmem_int, &sub_ivshmem_cmds,
```
I recommend using the macro `SHELL_CMD_ARG_REGISTER`:

```
SHELL_COND_CMD_ARG_REGISTER(ivshmem_int, &sub_ivshmem_cmds,
"Notify a vector on an ivshmem peer", cmd_ivshmem_int, 3, 0);
```

It will ensure that command is called with the proper number of parameters: `peer_id` and `vector`.

tbursztyka replied:
true

nashif commented on drivers/CMakeLists.txt:
```diff
@@ -41,6 +41,7 @@ add_subdirectory_ifdef(CONFIG_LORA lora)
 add_subdirectory_ifdef(CONFIG_PECI peci)
 add_subdirectory_ifdef(CONFIG_REGULATOR regulator)
 add_subdirectory_ifdef(CONFIG_MEMC memc)
+add_subdirectory_ifdef(CONFIG_MISC misc)
```
CONFIG_MISC is really a wildcard here and can be anything. We should find an existing category or create a new one. Maybe ipm? or ipc? or even shared_memory, anything, but not MISC please :-)

tbursztyka replied:
I was out of inspiration, I admit. virtualization? guest? Maybe the first. I wouldn't be surprised to see more virtualization/emulation stuff landing in zephyr at some point.

tbursztyka replied:
made it to virtualization

jakub-uC commented on drivers/misc/misc_ivshmem_shell.c:
```diff
@@ -0,0 +1,180 @@
+
+SHELL_CMD_ARG_REGISTER(ivshmem_int, &sub_ivshmem_cmds,
+		       "Notify a vector on an ivshmem peer",
+		       cmd_ivshmem_int, 2, 0);
```
It shall be `3`, not `2`.
```
argv[0] == "ivshmem_int"
argv[1] => peer_id 
argv[2] => vector 
```

tbursztyka replied:
Ok, the documentation is a bit unclear at is talks about mandatory/optional arguments, which I understood as arguments _to_ the command. (I did not see the command itself would be part of the arguments, maybe I missed it)

jakub-uC replied:
Fair point. I will update the documentation

jakub-uC commented on drivers/virtualization/virt_ivshmem_shell.c:
```diff
@@ -0,0 +1,180 @@
+		       "Notify a vector on an ivshmem peer",
+		       cmd_ivshmem_int, 3, 0);
+
+SHELL_CMD_ARG_REGISTER(ivshmem_get_notified, &sub_ivshmem_cmds,
```
Just a proposal.
You can remove `ifdef` from the command handler and define command like that:
```
SHELL_COND_CMD_ARG_REGISTER(CONFIG_IVSHMEM_DOORBELL,
                       ivshmem_get_notified, &sub_ivshmem_cmds,
                       "Get notification on vector",
		       cmd_ivshmem_get_notified, 2, 0);
```

The same goes for the subcommand `get_notified`. There is a macro: `SHELL_COND_CMD_ARG`

tbursztyka replied:
I'll let it as it is so the user will know if the doorbell feature is enabled or not.

tbursztyka commented:
@jakub-uC I have an issue with the notification thread and printing to shell (on ivshmem_get_notified). It crashes the system. I guess there are rights issues. If I comment out the shell_fprintf() there, it's fine. Any idea which rights should I grant to the thread?

nashif commented:
you need to rebase

jakub-uC reviewed (approved):
LGTM from the shell perspective

tbursztyka commented:
Adding DNM label: I am adding documentation and, if possible, hooks to start qemu with ivshmem (plain or doorbell) parameters included.

Commit:
doc: Adding ivshmem general and API documentation

A quick summary on what is ivshmem and its Zephyr driver API.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

jukkar commented on include/drivers/virtualization/ivshmem.h:
```diff
@@ -0,0 +1,144 @@
+extern "C" {
+#endif
+
+typedef size_t (*ivshmem_get_mem_f)(const struct device *dev,
```
Just wondering about the typedefs and naming. At least in networking side, the callback typedefs have been called `..._cb_t`, is this a new zephyr convention to use `.._f`?

tbursztyka replied:
..._cb_t would be meaningful for an actual callback, ivshmem does not expose any.
That said there is no really naming convention on these specific types (which are only made to describe the API structure attributes). Some do not put anything at the end, some put _fn, some _f...

Commit:
drivers/virtualization: Make ivshmem driver userspace ready

Exposing its public API to userspace.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
cmake: Adds the necessary flags support for Qemu ivshmem feature

Depending whether doorbell or plain versions are enabled, it will set
the right flags to qemu.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
tests/boards: Add a test for ivshmem on qemu_x86_64

It runs on userspace directly. Only the ivshmem-plain version is tested,
as is not possible to run other VM and tools in parallel.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
DNM label removed as I added documentation, cmake hooks for setting relevant qemu flags and a unit-test.

tbursztyka commented:
@jhedberg @jukkar another round?
Btw, how can I trigger documentation check? (there is only 3 checks, where usually it's up to 7...)

Commit:
CODEOWNERS: Adding /drivers/virtualization owner

Adding myself there.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

nashif commented:
i have no idea why this PR keeps getting stuck!

tbursztyka commented:
@nashif recent change of memory management made the ivshmem driver broken. s/z_mem_map/z_phys_map. It should be fine now

nashif commented:
>  It should be fine now

doc build broke :)

tbursztyka commented:
@nashif the url format is a bit weird, looks like a '_' is required after the `alias <url>`... let's see
